Source/Core/DolphinQt/GameList/GameTracker.cpp

```diff
@@ -46,92 +48,76 @@ GameTracker::GameTracker(QObject* parent) : QFileSystemWatcher(parent)
   });
 
   connect(&Settings::Instance(), &Settings::MetadataRefreshRequested, [this] {
-    m_load_thread.EmplaceItem(Command{CommandType::UpdateMetadata, {}});
+    m_cache.UpdateAdditionalMetadata(
+        [this](const std::shared_ptr<const UICommon::GameFile>& game) { emit GameUpdated(game); },
+        m_aborted);
+    Settings::Instance().NotifyMetadataRefreshComplete();
   });
 
-  m_load_thread.Reset([this](Command command) {
-    switch (command.type)
-    {
-    case CommandType::LoadCache:
-      LoadCache();
-      break;
-    case CommandType::Start:
-      StartInternal();
-      break;
-    case CommandType::AddDirectory:
-      AddDirectoryInternal(command.path);
-      break;
-    case CommandType::RemoveDirectory:
-      RemoveDirectoryInternal(command.path);
-      break;
-    case CommandType::UpdateDirectory:
-      UpdateDirectoryInternal(command.path);
-      break;
-    case CommandType::UpdateFile:
-      UpdateFileInternal(command.path);
-      break;
-    case CommandType::UpdateMetadata:
-      m_cache.UpdateAdditionalMetadata(
-          [this](const std::shared_ptr<const UICommon::GameFile>& game) {
-            emit GameUpdated(game);
-          });
-      QueueOnObject(this, [] { Settings::Instance().NotifyMetadataRefreshComplete(); });
-      break;
-    case CommandType::PurgeCache:
-      m_cache.Clear(UICommon::GameFileCache::DeleteOnDisk::Yes);
-      break;
-    case CommandType::BeginRefresh:
-      if (m_busy_count++ == 0)
-      {
-        for (auto& file : m_tracked_files.keys())
-          emit GameRemoved(file.toStdString());
-        m_tracked_files.clear();
-      }
-      break;
-    case CommandType::EndRefresh:
-      if (--m_busy_count == 0)
-      {
-        QueueOnObject(this, [] { Settings::Instance().NotifyRefreshGameListComplete(); });
-      }
-      break;
-    }
-  });
+  QThread::start(QThread::IdlePriority);
 
-  m_load_thread.EmplaceItem(Command{CommandType::LoadCache, {}});
+  LoadCache();
 
   // TODO: When language changes, reload m_title_database and call m_cache.UpdateAdditionalMetadata
 }
 
-void GameTracker::LoadCache()
+GameTracker::~GameTracker()
 {
-  m_cache.Load();
-  m_cache_loaded_event.Set();
+  m_aborted = true;
+  QThread::quit();
+  const bool joined = QThread::wait(5000);
```

Comment:
This timeout seems sloppy. In testing it does time out though. Why?
The GameTracker thread was terminated within `UpdateDirectory`.
I didn't dig much deeper but maybe more functions need to be aware of the abort variable?

Reply:
I've now added a check in `UpdateDirectory()` to bail out if aborted. Also, for solving the stall while purging, I added a separate `m_pendingTask` member, that will be set when purge or refresh all are requested; this guarantees the loop in `UpdateDirectory()` is halted.